python-flask/app/__init__.py

```diff
@@ -0,0 +1,42 @@
+from distutils.command.config import config
+
+from flask import Flask, session, g
+from app.extensions import db, migrate
+from app.main.routes import main_bp
+from app.posts.routes import posts_bp
+from app.auth.routes import admin_bp
+from app.simulator.routes import sim_bp
+from app.utils import format_datetime, gravatar
+from app.models.user import User
+
+def before_request():
+    """Runs before every request to check if a user is logged in."""
+    g.user = None
+    if 'user_id' in session:
+        g.user = User.query.filter_by(user_id=session['user_id']).first()
+
+def create_app():
+    app = Flask(__name__)
+
+    # Configurations
+    app.config.from_pyfile('../config.py')
+
+    # Initialize extensions
+    db.init_app(app)
+    migrate.init_app(app, db)
+
+    # Register Blueprints
+    app.register_blueprint(main_bp)
```

Comment:
maybe a function here only to register the blueprint would be nice. Then you call just one function in the main.

Reply:
I think the way I did it is the common way for the flaks framework. See documentation here:
https://realpython.com/flask-blueprint/

Reply:
I understand and it is correct man. It is a personal opinion but I would change this file to be something like this: 
```suggestion
from flask import Flask,
def create_app():
   app = Flask(__name__)
   server_init(app)
   return app
```

And then I would create a xxxxxx.py with all this structure. (Maybe put it in the same file also is an option). Something like this:
```suggestion
from distutils.command.config import config
from flask import Flask, session, g
from app.extensions import db
from app.main.routes import main_bp
from app.posts.routes import posts_bp
from app.auth.routes import admin_bp
from app.simulator.routes import sim_bp
from app.utils import format_datetime, gravatar
from app.models.user import User

def is_user_logged():
    """Runs before every request to check if a user is logged in."""
    g.user = None
    if 'user_id' in session:
        g.user = User.query.filter_by(user_id=session['user_id']).first()

def register_blueprints(app):
    app.register_blueprint(main_bp)
    app.register_blueprint(posts_bp)
    app.register_blueprint(admin_bp)
    app.register_blueprint(sim_bp, url_prefix='/api')
    
def configure_server(app):
   app.config.from_pyfile('../config.py')
   app.debug = app.config['DEBUG'] # Do we need this DEBUG?
   db.init_app(app) #here we are binding the SQLalchemy to flask right?
   app.before_request(is_user_logged)
   
def set_environment_filters(app):
    app.jinja_env.filters['format_datetime'] = format_datetime
    app.jinja_env.filters['gravatar'] = gravatar
    
def prepare_application(app):
   configure_server(app)
   register_blueprints(app)
   set_environment_filters(app)
   
```

I know that is very annoying and I'm sorry. I just think it is easier to understand what is going on this way and also leaves the __init__ pretty clean.
PS: I don't know if this code will work. Im just assuming it.

Reply:
Changed this! I created an `app_setup.py` file based on the suggestions. I discovered that when trying to run it locally, the `SQLALCHEMY_DATABASE_URI` is not loaded correctly. It only works if we pass the env variables in `docker-compose'. This problem seems to be related to the environment variable substitution in the .env file, which does not work when using variable interpolation as in shell scripts. We may need to change this back in the future to: 
```
POSTGRES_USER = os.getenv('POSTGRES_USER')
POSTGRES_PASSWORD = os.getenv('POSTGRES_PASSWORD')
POSTGRES_HOST = os.getenv('POSTGRES_HOST')
POSTGRES_PORT = os.getenv('POSTGRES_PORT')
POSTGRES_DB = os.getenv('POSTGRES_DB')

# Construct the full SQLALCHEMY_DATABASE_URI
SQLALCHEMY_DATABASE_URI = f"postgresql://{POSTGRES_USER}:{POSTGRES_PASSWORD}@{POSTGRES_HOST}:{POSTGRES_PORT}/{POSTGRES_DB}" 
```
Don't know what's the best solution tbo. Guess when we running it in docker it should be fine.

Reply:
Changed this! I created an `app_setup.py` file based on the suggestions. I discovered that when trying to run it locally, the `SQLALCHEMY_DATABASE_URI` is not loaded correctly. It only works if we pass the env variables in `docker-compose'. This problem seems to be related to the environment variable substitution in the .env file, which does not work when using variable interpolation as in shell scripts. We may need to change this back in the future to: 
```
POSTGRES_USER = os.getenv('POSTGRES_USER')
POSTGRES_PASSWORD = os.getenv('POSTGRES_PASSWORD')
POSTGRES_HOST = os.getenv('POSTGRES_HOST')
POSTGRES_PORT = os.getenv('POSTGRES_PORT')
POSTGRES_DB = os.getenv('POSTGRES_DB')

# Construct the full SQLALCHEMY_DATABASE_URI
SQLALCHEMY_DATABASE_URI = f"postgresql://{POSTGRES_USER}:{POSTGRES_PASSWORD}@{POSTGRES_HOST}:{POSTGRES_PORT}/{POSTGRES_DB}" 
```
Don't know what's the best solution tbo. Guess when we running it in docker it should be fine.